### PR TITLE
Runtime and Many New Features Support

### LINKED ISSUE #6: Default Reporter should pipe log values to stdout instead of console.log()

### LINKED ISSUE #16: Add toInclude and toIncludeEqual

### LINKED ISSUE #52: [Bug] .npmignore not including certain files
# Files Not Properly Ignored by `.npmignore`

- `__tests__`
- `assembly/__tests__` (don't forget to *include* the types file)
- `jest.config.js`

### LINKED ISSUE #56: Feature requests: filtering tests by name
It would be nice if `as-pect` can filter tests by name like Rust, although since tests usually only have description, we might want to consider some other way of filtering.

### LINKED ISSUE #58: [Feature] --no-run and --output-binary flag
We need a way to capture the output of the compiler so that binaries can be saved and run later, say, from a web server.

### LINKED ISSUE #60: [Feature] Support Compiler Flags From Command Line

### LINKED ISSUE #61: [Feature] Support AssemblyScript Runtime
This is already in progress. There are some issues with the way references are compared and they need to be refractored.

### LINKED ISSUE #63: [Feature] CSV/JSON reporter
This is good for test reports.  For example, when grading students it would be handy to produce a CSV with each group summarized with number of passing tests. e.g.

```
2,1,10
```

### LINKED ISSUE #73: Reporting Expected and Actual `u64` and `i64` values doesn't work
Any advice on the matter @willemneal ?

### PR DESCRIPTION
**Pull request by jtenner**
closes #61 
fixes #53 
fixes #52 
closes #56 
closes #58 
closes #16 
closes #60 
closes #64 
closes #63 
closes #6 
fixes #73

### CODE REVIEW

File: assembly/__tests__/setup/Vec3.ts [hunk 1]
```diff
@@ -13,31 +13,27 @@ export class Vec3 {
...
 
-  @inline
   @operator("<")
   protected __lessThan(reference: Vec3): bool {
     var magnitude = this.magnitude();
     var refmagnitude = reference.magnitude();
     return magnitude < refmagnitude;
   }
 
-  @inline
```
Comment (willemneal) [inline assembly/__tests__/setup/Vec3.ts]: Just curious why this change?

**Comment by jtenner** [comment]
The assemblyscript compiler emits a warning when it tries to inline an
operator like this. When it becomes supported I will add it back.

File: assembly/__tests__/setup/Vec3.ts (hunk 1 above)
Comment (jtenner) [reply assembly/__tests__/setup/Vec3.ts]: The assemblyscript compiler emits a warning when it tries to inline an
operator like this. When it becomes supported I will add it back.

**Comment by jtenner** [comment]
The branch now no longer has 100% TestContext line coverage.

We will need to test the `testRegex` and `groupRegex` matchers.  This is very easy, but requires an additional binary compilation when jest runs. Will work on this tomorrow.

**Comment by jtenner** [comment]
Coverage is now at 100%. Halfway finished with `--output-binary` and `--no-run`. While, `--output-binary` is completed, `--no-run` requires that no output is pushed to the console about the tests. Will need some more time to figure out how this flag works.

**Comment by jtenner** [comment]
The `--norun` flag is now added. `--no-run` is parsed by `yargs` as `run: false` instead of `noRun: true`.

**Comment by jtenner** [comment]
The next feature that needs to be added is compiler flags from the command line. 

This is going to require a huge cli parser refactor and should probably ship with `2.0.0`. Do you have any objections @willemneal. This can include #60  and #29.

If this is the case, all we need to do is add a new assertion called `toInclude` and `toIncludeEqual` in the assertions.

```ts
const array: number[] = [1, 2, 3];
expect(array).toInclude(2, "array should have 2");
```

**Comment by willemneal** [comment]
Yeah I think it's fine to wait on the command line args.  Though I don't think it'll require as much rewriting as you suggest.  The way that npm lets you pass arguments to scripts is by using `--`  So we just have to split over the ` -- ` and then pass the first half to yargs.

**Comment by jtenner** [comment]
Okay. I added toInclude and will implement  toIncludeEqual.

If you want, @willemneal you can implement command line args.

**Comment by jtenner** [comment]
The only problem I see with that is that command line args should overwrite compiler flags passed via configuration

**Comment by jtenner** [comment]
Is there any good way to override compiler flags provided by the configuration?

File: assembly/__tests__/setup/Vec3.ts (hunk 1 above)
Comment (jtenner) [reply assembly/__tests__/setup/Vec3.ts]: All set? I don't want to mark this conversation as resolved without running it by you first.

**Comment by willemneal** [comment]
I'll make a fork of the compiler flags and add the overwritting.  I think the easiest way is to use Object.assign, which will use the compiler flags passed first and the config file flags won't overwrite them.

**Comment by jtenner** [comment]
Yeah there's only one problem. There are some flags you can pass multiple times. Is there a way to handle these?

File: src/test/TestContext.ts [hunk 2]
```diff
@@ -1,139 +1,100 @@
...
+    this.startupTime = timeDifference(performance.now(), start);
+
+    if (this.errors.length > 0) return;
+    this.ready = true;
 
     // start the test suite
     this.reporter.onStart(this);
-    runContext.start = performance.now();
 
     testgroup:
```
Comment (willemneal) [inline src/test/TestContext.ts]: This is never referenced.

**Comment by willemneal** [comment]
@jtenner Since we're using the same argument parser as asc, then it will be same as if they passed multiple arguments manually to `asc.`  Not sure what would happen, but that's outside the scope of this project.

File: src/test/TestContext.ts (hunk 2 above)
Comment (jtenner) [reply src/test/TestContext.ts]: What's not referenced? The `start` and `end` values used by `performance.now()` aren't stored on the `TestContext`.

It's just used for calculating the `.time` value.

Comment (jtenner) [reply src/test/TestContext.ts]: I removed this label

**Comment by jtenner** [comment]
I'm glad I took the time to generate this pull request. It contains a lot of bug fixes and features that I think are necessary to make `as-pect` enterprise-grade testing software.

I think I need to spend some more time with the documentation in order to make this pull request more complete.

**Comment by jtenner** [comment]
I think this is now *very* stable, but I would love to release this as a beta. This is a backward incompatible rewrite because of all the memory layout changes, so it will be released as `1.0.0-beta`.

@willemneal I think all this source code could use a huge code review when you get the chance. There are lots to optimize and things to comment. Please help me point them out!

**Comment by willemneal** [comment]
Okay great!  I just tested this with my new compiler frontend and am getting a runtime of 4.2 s.  Should I make a PR to this branch to add my assemblyscript as the dependency and change the way it's imported.

**Comment by jtenner** [comment]
Yes please. Make a fork so I can review it. I want to make sure I understand what you did.

**Comment by jtenner** [comment]
I need to make one more commit.

File: assembly/__tests__/toInclude.spec.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
+import { Vec3 } from "./setup/Vec3";
+
+let numberTester: i32[] = new Array<i32>();
```
Comment (willemneal) [inline assembly/__tests__/toInclude.spec.ts]: Shouldn't this be in a `beforeAll` block?

File: assembly/internal/comparison/arrayComparison.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
+import { assert } from "./assert";
+import { reportActual } from "../report/reportActual";
+import { reportExpected } from "../report/reportExpected";
+
+// @ts-ignore: Decorators *are* valid here
+@inline
+export function arrayComparison<T>(actual: T, expected: T, negated: i32, message: string): void {
```
Comment (willemneal) [inline assembly/internal/comparison/arrayComparison.ts]: Is `<T extends Array>` possible?  If not it would be a good feature request for upstream.

File: assembly/internal/comparison/closeToComparison.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
+import { reportActual } from "../report/reportActual";
+import { reportExpected } from "../report/reportExpected";
+import { assert } from "./assert";
+
+// @ts-ignore: Decorators *are* valid here!
+@inline
+export function closeToComparison<T>(actual: T, expected: T, negated: i32, decimalPlaces: i32, message: string): void {
```
Comment (willemneal) [inline assembly/internal/comparison/closeToComparison.ts]: Again it would be awesome if we could do `<T extends float>`, where float is `type float = f32 | f64;`  But again this feature should be pushed upstream.

Comment (jtenner) [reply assembly/internal/comparison/closeToComparison.ts]: I like the idea that this assertion is a runtime error.

```ts
assert(isFloat<T>(), "Message");
```

This is good enough for now.

File: assembly/internal/comparison/arrayComparison.ts (hunk 4 above)
Comment (jtenner) [reply assembly/internal/comparison/arrayComparison.ts]: Yes. This is not possible. `Array` has a generic `T` type parameter that must be provided, and we cannot know what `U` is up front. This is why `Array<U>` where `U` is a reference type is not supported.

File: assembly/__tests__/toInclude.spec.ts (hunk 3 above)
Comment (jtenner) [reply assembly/__tests__/toInclude.spec.ts]: This is a global value, and the value is created once.

If I want to add a future `describe()` block, it will add those items each time a `describe()` block is run.

I only want it to run once.

File: assembly/internal/report/reportExpected.ts [hunk 6]
```diff
@@ -0,0 +1,171 @@
+import { ValueType } from "./ValueType";
```
Comment (willemneal) [inline assembly/internal/report/reportExpected.ts]: The reporter files are certainly improved, but still feels like there is code duplication going one.  Couldn't they be combined like `reportFloat(negated: i32, expect: f64, actual: f64)`.  If separating them makes sense then ignore me.

File: assembly/internal/comparison/closeToComparison.ts (hunk 5 above)
Comment (willemneal) [reply assembly/internal/comparison/closeToComparison.ts]: True, just saying that we can reduce runtime checks that should be compile time errors.

File: assembly/internal/report/reportExpected.ts (hunk 6 above)
Comment (jtenner) [reply assembly/internal/report/reportExpected.ts]: This is because reporting a `float`, `i32`, or `u32` doesn't require `long` support. Reporting an `i64` or a `u64` is a special case that requires some JavaScript hackery.

Comment (willemneal) [reply assembly/internal/report/reportExpected.ts]: I meant in the more general case; having a single report function versus having two.

Comment (jtenner) [reply assembly/internal/report/reportExpected.ts]: Also, it uses the same external function and simply provides a different signature for convenience. Generic `T` types might be supported by AssemblyScript for linked functions like this, but this makes things very clear about what is going on.

Just about the only argument to be made for using the "same" function with a generic parameter is speed of execution, which is negligable, since the compiler would need to compile two versions of the function anyway.

File: assembly/internal/report/reportExpected.ts [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+@external("__aspect", "reportExpectedValue")
+declare function reportExpectedInteger(value: i32, negated: i32): void;
+
+// @ts-ignore: Decorators *are* valid here!
+@external("__aspect", "reportExpectedReference")
+declare function reportExpectedReferenceExternal(value: usize, offset: i32, negated: i32): void;
+
+// @ts-ignore: Decorators *are* valid here!
+@external("__aspect", "reportExpectedString")
+declare function reportExpectedString(value: string, negated: i32): void;
```
Comment (willemneal) [inline assembly/internal/report/reportExpected.ts]: What I'm trying to say is why not have `reportString(expect: string, actual: string, negated: i32): void`, etc, instead of breaking it into two calls?

File: assembly/internal/comparison/closeToComparison.ts (hunk 5 above)
Comment (jtenner) [reply assembly/internal/comparison/closeToComparison.ts]: This is incorrect. We don't know what `T` is at compile time until the function is called. 

`T` itself is a parameter of the `expect<T>()` function, thus, we cannot assert that `T` is a float without a runtime check.

File: assembly/internal/report/reportExpected.ts (hunk 7 above)
Comment (jtenner) [reply assembly/internal/report/reportExpected.ts]: I understand now.

This was a personal development choice I made because sometimes we compare a float value to a string (a la `toBeFinite()` which compares against the string "finite".) Leaving things separate makes things clearer and does not make things too much slower. Remember that test files of reasonable size run in `25ms`. This has "premature" optimization written all over it.

File: assembly/internal/comparison/closeToComparison.ts (hunk 5 above)
Comment (willemneal) [reply assembly/internal/comparison/closeToComparison.ts]: Reading this I thought that `expect<T>(x).toBeCloseTo` means that the type parameter would be passed to closeToComparison<T>.  Thus at compile time the type could be checked to make sure it extends float (that is if that is possible in the first place).

```ts
 @inline
  public toBeCloseTo(expected: T, decimalPlaces: i32 = 2, message: string = ""): void {
    closeToComparison<T>(this.actual, expected, this._not, decimalPlaces, message);
  }
```

Comment (jtenner) [reply assembly/internal/comparison/closeToComparison.ts]: We don't know if `T` is an `f32` or an `f64` at compile time. We can only pass the type along to the comparison.

**Comment by willemneal** [comment]
What is reasonable size?  And regardless all I'm saying is that these seem
like duplicated code. Which makes it feel like something is not right. I'm
fine as is just saying easy to develop at first doesn't always mean what it
should always be.

File: assembly/internal/comparison/closeToComparison.ts (hunk 5 above)
Comment (jtenner) [reply assembly/internal/comparison/closeToComparison.ts]: Also, that function should not be inlined.

File: assembly/internal/report/reportExpected.ts (hunk 7 above)
Comment (jtenner) [reply assembly/internal/report/reportExpected.ts]: Modules of reasonable size include a test file of at least 50 tests.

For instance the `toStrictEqual()` tests run in less than `8ms` despite the fact that there are 25 tests. Asking for more performance out of these functions will produce a very minute performance gain, especially since they are only called when a test fails.

File: assembly/internal/comparison/closeToComparison.ts (hunk 5 above)
Comment (willemneal) [reply assembly/internal/comparison/closeToComparison.ts]:
```
  it("should expect values that are close", (): void => {
    expect<f64>(1.0).toBeCloseTo(1.00001, 2, "1.000001 is close to 1.0");
  });
```
The compiler does know what `T` is.  My point was if we had a way to type check that the generic type was either floats then it would save the dynamic check.  It's my complaints with AS, not your code.

File: assembly/internal/report/reportExpected.ts (hunk 7 above)
Comment (willemneal) [reply assembly/internal/report/reportExpected.ts]: For me it's not the efficiency, but rather the design choice, which like I said if it was put in place for a reason and changing would make things harder to work with don't bother.  Rather trying to understand the design choice of two files that seem almost identical.

**Comment by willemneal** [comment]
LGTM!  Just my question about design, but those are just for my own understanding and gripes about the current state of assembyscript.

File: assembly/internal/report/reportExpected.ts (hunk 7 above)
Comment (jtenner) [reply assembly/internal/report/reportExpected.ts]: I agree the files are identical and it feels like the code is repeated. I could be convinced to combine these functions together, but these classes are essentially static global variables that perform two slightly different explicit operations. If you can help me design something that is much clearer and makes things easy to work with, I will change it. Definitely.

In the meantime, this is the best compromise I could personally come up with.

**Comment by jtenner** [comment]
@willemneal your experience is really valuable and I appreciate all your effort. When you think this is ready, I will begin the process of documenting each function, and its purpose.

File: assembly/internal/comparison/closeToComparison.ts (hunk 5 above)
Comment (jtenner) [reply assembly/internal/comparison/closeToComparison.ts]: Agreed.  I don't think there's an option to make this check happen at compile time.

**Comment by jtenner** [comment]
@willemneal I increased the documentation by a lot and fixed signed integer logging and reporting. I thought it was weird when I saw `4294967294` get reported as `-2`.

One last change needs to be made because the `@start` decorator is going to be removed in favor of the `--explicitStart` flag.

**Comment by jtenner** [comment]
@willemneal Can you approve the request? I did touch almost every single file in the project now.  Now that assemblyscript master has been updated, I'd like to finish this pull request.
